proposed/2023/NuGetAudit-without-nuget.org-as-packageSource.md

```diff
@@ -0,0 +1,217 @@
+<!-- Are there any resources that are relevant to this proposal? -->
+Although [`npm audit`'s help docs](https://docs.npmjs.com/cli/v10/commands/npm-audit?v=true) don't mention it, it appears to have a `--registry={url}` argument, as can be seen on [the Azure DevOps docs on npm audit](https://learn.microsoft.com/azure/devops/artifacts/npm/npm-audit?view=azure-devops&tabs=yaml).
+
+Pypi's `pip` has [an audit command](https://pypi.org/project/pip-audit/), and `--index-url` and `--extra-index-url` appear to enable using custom audit sources.
+
+Rust's cargo audit command has [configuration for the advisory DB URL](https://docs.rs/cargo-audit/latest/cargo_audit/config/struct.DatabaseConfig.html)
+
+Note that for these other package ecosystems, they do not appear to need any explicit configuration file in order to specify default sources (either package source, or audit source).
+Instead, defaults appear hard-coded into the app, and it requires explicit configuration to remove.
+
+Contrast this to NuGet's design where NuGet creates a default configuration file at first use, and then requires sources to be explicitly defined.
+However, NuGet settings that fall under the `<config>` section of the *nuget.config* file do have hard-coded defaults that do not need to be explicitly defined in any config file.
+
+Both choices have advantages and disadvantages.
+Needing explicit config files, as NuGet's package sources do, means it might be easier for customers to discover why a source is being used.
+Also, customers are less likely to have unexpected changes if the app changes defaults.
+On the other hand, when customers are happy to change with changing defaults, having an explicit config file makes it much more difficult for customers to learn about changed defaults, and from the package manager team's point of view, much more difficult to have customers adopt new defaults.
+
+## Unresolved Questions
+
+<!-- What parts of the proposal do you expect to resolve before this gets accepted? -->
+<!-- What parts of the proposal need to be resolved before the proposal is stabilized? -->
+<!-- What related issues would you consider out of scope for this proposal but can be addressed in the future? -->
+
+All resolved during the spec review process.
+
+## Future Possibilities
+
+<!-- What future possibilities can you think of that this proposal would help with? -->
+
+### Allow auditSources to point to a filesystem directory
+
+In order to provide customers who are disconnected from nuget.org (and possibly the internet more widely), we could load the VDB from files on disk.
+It would be up to customers to determine how to copy the VDB to their disconnected network/machine.
+We could consider a .NET tool, for example `nuget-audit-download` which knows how to communicate with nuget.org's V3 protocol, and merge results into a single file, so that it's easy for customers to obtain the file that needs to be copied.
```

Comment:
We don't need the .NET tool for this section, right? It would just be like adding any folder feed, so nuget would utilize that feed to find the VDB?
You're just saying a secondary tool could be developed outside of nuget.exe/dotnet nuget?

Reply:
Correct. The only way today for customers to download NuGet's VDB is to understand our server API, then either develop a tool themselves, or open https://api.nuget.org/v3/index.json in their web browser, and follow the links to the VDB files and then save them.

What I'm implying with the last sentence is that if we think that customers who can't use nuget.org as an audit source is a common enough scenario for us to support officially, then we could provide a tool to make this as easy as possible for them. But there's no reason we have to. We could leave it for the community to develop themselves.